src/geocat/comp/climatology.py

```diff
@@ -322,3 +352,140 @@ def month_to_season(
     # Therefore, we must filter out the NaNs
     return means.sel(
         {time_coord_name: means[time_coord_name].dt.month == months[1]})
+
+
+def calendar_average(
+        dset: typing.Union[xr.DataArray, xr.Dataset],
+        freq: str,
+        time_dim: str = None,
+        climatology: bool = False,
+        calendar: str = None) -> typing.Union[xr.DataArray, xr.Dataset]:
+    """This function computes averages according to a given time frequency.
+
+    Parameters
+    ----------
+    dset : :class:`xarray.Dataset`, :class:`xarray.DataArray`
+        The data on which to operate. It must be evenly spaced in the time dimension.
+
+    freq : :class:`str`
+        Frequency alias. Accepted alias:
+            - 'hour': for hourly averages
+            - 'day': for daily averages
+            - 'month': for monthly averages
+            - 'season': for meteorological seasonal averages (DJF, MAM, JJA, and SON)
+            - 'yearly': for yearly averages
+
+    time_dim : :class:`str`, Optional
+        Name of the time coordinate for `xarray` objects
+
+    climatology : :class:`bool`, Optional
+        Default False. If False, the average for each period (day, month, etc.)
+        will be calculated for it's given year (i.e. the average for Jan-2000
+        will be independent of the average for Jan-2001). If True,
+        climatological averages will be calculated across all years provided.
+
+    calendar : :class:`str`, Optional
+        Alias for the kind of calendar the time dimension of the data is in.
+        Defaults to `None`, in which case the calendar is infered from the data.
+        A list of accepted aliases can be found `here <http://xarray.pydata.org/en/stable/generated/xarray.cftime_range.html>`_
+
+    Returns
+    -------
+    computed_dset: same type as dset
+        The computed data
+
+    Notes
+    -----
+    Seasonal averages are weighted based on the number of days in each month.
+    This means that the given data must be monotonic (i.e. data every 6 hours,
+    every two days, every month, etc.) and must not cross month boundaries
+    (i.e. don't use weekly averages where the week falls in two
+    different months)
+    """
+    # TODO: add functionality for users to select specific seasons or hours for avgs/clims
+    freq_dict = {
+        'hour': ('%m-%d %H', 'H', '30min'),
+        'day': ('%m-%d', 'D', '12H'),
+        'month': ('%m', 'MS', 'SMS'),
+        'season': (None, 'QS-DEC', 'MS'),
+        'year': (None, 'YS', '6MS')
+    }
+
+    try:
+        freq_dict[freq]
+    except KeyError:
+        raise KeyError(
+            f"contributed: calendar_average: bad period: PERIOD = {freq}. Valid periods include: {list(freq_dict.keys())}"
+        )
+
```

Comment:
```suggestion
    if freq not in freq_dict:
    	raise KeyError(
            f"Received bad period {freq!r}. Expected one of {list(freq_dict.keys())!r}"
        )

```

This is the canonical python pattern. I reworded the message to be clearer

Reply:
I wasn't aware of this unspoken standard for error messages. Thanks!

Reply:
SOrry that wasn't clear. The standard is the `if freq  not in freq_dict` check as opposed to checking for a `KeyError`. I just like that format of error message. It's very clear!